js/compiler.ts

```diff
@@ -0,0 +1,502 @@
+// Copyright 2018 the Deno authors. All rights reserved. MIT license.
+import * as ts from "typescript";
+import { assetSourceCode } from "./assets";
+import * as deno from "./deno";
+import { libdeno, window, globalEval } from "./globals";
+import * as os from "./os";
+import { RawSourceMap } from "./types";
+import { assert, log } from "./util";
+import * as sourceMaps from "./v8_source_maps";
+
+const EOL = "\n";
+const ASSETS = "$asset$";
+
+// tslint:disable-next-line:no-any
+type AmdFactory = (...args: any[]) => object | undefined;
+type AmdDefine = (deps: string[], factory: AmdFactory) => void;
+
+declare global {
+  interface Window {
+    define?: AmdDefine;
+  }
+}
+
+/**
+ * Abstraction of the APIs required from the `os` module so they can be
+ * easily substituted.
+ */
+export interface Os {
+  codeCache: typeof os.codeCache;
+  codeFetch: typeof os.codeFetch;
+  exit: typeof os.exit;
+}
+
+/**
+ * Abstraction of the APIs required from the `typescript` module so they can
+ * be easily substituted.
+ */
+export interface Ts {
+  createLanguageService: typeof ts.createLanguageService;
+  /* tslint:disable-next-line:max-line-length */
+  formatDiagnosticsWithColorAndContext: typeof ts.formatDiagnosticsWithColorAndContext;
+}
+
+/**
+ * A simple object structure for caching resolved modules and their contents.
+ *
+ * Named `FileModule` to clarify it is just a representation of meta data of
+ * the module, not the actual module instance.
+ */
+class FileModule {
+  scriptVersion = "";
+  readonly exports = {};
+
+  constructor(
+    public readonly fileName: string,
+    public readonly sourceCode = "",
+    public outputCode = ""
+  ) {
+    if (outputCode !== "") {
+      this.scriptVersion = "1";
+    }
+  }
+}
+
+/**
+ * The required minimal API to allow formatting of TypeScript compiler
+ * diagnostics.
+ */
+const formatDiagnosticsHost = {
```

Comment:
Is there an interface that this implements? If so it would be good to add a type here.